File: todo-management/src/main/java/net/samik/todo/service/TodoService.java
package net.samik.todo.service;

import net.samik.todo.dto.TodoDto;
import java.util.*;

public interface TodoService {

    // Build Add Todo REST API
    TodoDto addTodo(TodoDto todoDto);

    // Build Get Todo REST API
    TodoDto getTodo(Long id);

    // Build Get All Todos REST API
    List<TodoDto> getAllTodos();

    // Build Update Todo REST API
    TodoDto updateTodo(Long id, TodoDto todoDto);

    // Build Delete Todo REST API
    void deleteTodo(Long id);

    // Build Complete Todo REST API
    TodoDto completeTodo(Long id);

    // Build InComplete Todo REST API
    TodoDto incompleteTodo(Long id);
}

File: todo-ui/src/services/TodoService.js
import React from "react";
import axios from "axios";
import { getToken } from "./AuthService";

const BASE_REST_API_URL = "http://localhost:8080/api/todos";

// Add a request interceptor
axios.interceptors.request.use(
  function (config) {
    config.headers["Authorization"] = getToken();
    return config;
  },
  function (error) {
    // Do something with request error
    return Promise.reject(error);
  }
);

const getAllTodos = () => {
  return axios.get(BASE_REST_API_URL);
};

const addTodo = (todo) => {
  return axios.post(BASE_REST_API_URL, {
    title: todo.title,
    description: todo.description,
    completed: todo.completed,
  });
};

const getTodo = (todoId) => {
  return axios.get(BASE_REST_API_URL + "/" + todoId);
};

const updateTodo = (todoId, todo) => {
  return axios.put(BASE_REST_API_URL + "/" + todoId, {
    title: todo.title,
    description: todo.description,
    completed: todo.completed,
  });
};

const deleteTodo = (todoId) => {
  return axios.delete(BASE_REST_API_URL + "/" + todoId);
};

const completeTodo = (todoId) => {
  return axios.patch(BASE_REST_API_URL + "/" + todoId + "/complete");
};

const inCompleteTodo = (todoId) => {
  return axios.patch(BASE_REST_API_URL + "/" + todoId + "/in-complete");
};

export {
  getAllTodos,
  addTodo,
  getTodo,
  updateTodo,
  deleteTodo,
  completeTodo,
  inCompleteTodo,
};
